core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go

```diff
@@ -227,15 +227,20 @@ func (bs *BlockSubscriber) processHead(h *evmtypes.Head) {
 	// head parent is a linked list with EVM finality depth
 	// when re-org happens, new heads will have pointers to the new blocks
 	i := int64(0)
-	for cp := h; cp != nil; cp = cp.Parent {
-		if cp != h && bs.blocks[cp.Number] != cp.Hash.Hex() {
-			bs.lggr.Warnf("overriding block %d old hash %s with new hash %s due to re-org", cp.Number, bs.blocks[cp.Number], cp.Hash.Hex())
+	cp := new(evmtypes.Head)
+	*cp = *h
+	for ; ; cp = cp.Parent {
+		if cp.Number != h.Number && bs.blocks[cp.Number] != cp.Hash.Hex() {
```

Comment:
I'm not sure i fully follow the cp.Number != h.Number logic here

Reply:
Am i correctly understanding that this can happen due to two reasons
1. The block was not existing and we are seeing it for the first time
2. The block existed before and got reorged

Reply:
`cp.Number != h.Number` makes sure that we don't log this for a proper new head.
your understanding is correct.
i do observe that there are few re-orgs but most cases are missing heads. e.g. first head is 100, next head is 103. hence. so yes, we can log differently.

Reply:
I think this overall loop can be a lot cleaner with something like this

```
for ;cp !=nil ; cp = cp.Parent {
if bs.blocks[cp.Number] == cp.Hash.Hex() {
   break
}
if (bs.blocks[cp.Number] == zeroHex) {
   log("new block found")
} else {
   log("block reorged")
}
bs.blocks[cp.Number] = cp.Hash.Hex()
i++
if i > bs.blockSize {
 	break
 }
```

wydt?